openvdb_houdini/Utils.h

```diff
@@ -339,6 +339,68 @@ using VolumeGridTypes = ScalarGridTypes::Append<Vec3GridTypes>;
 
 using AllGridTypes = VolumeGridTypes::Append<PointGridTypes>;
 
+
+namespace internal {
+
+// Functor for use with GEOvdbApply() to deep copy a grid's tree,
+// if it is shared with other grids, before invoking a user-supplied
+// functor on the grid
+/// @private
+template<typename OpT>
+class MakeUniqueOp
+{
+public:
+    explicit MakeUniqueOp(OpT& op): mOp(op) {}
+
+    template<typename GridT>
+    void operator()(GridT& grid)
+    {
+        auto treePtr = grid.baseTreePtr();
+        if (treePtr.use_count() > 2) { // grid + treePtr = 2
+            grid.setTree(treePtr->copy());
+        }
+        mOp(grid);
+    }
+
+private:
+    OpT& mOp;
+};
+
+} // namespace internal
+
+
+/// @brief If the given primitive's grid resolves to one of the listed grid types,
+/// invoke the functor @a op on the resolved grid.
+/// @return @c true if the functor was invoked, @c false otherwise
+template<typename GridTypeListT, typename OpT>
+inline bool
+GEOvdbApply(const GEO_PrimVDB& vdb, OpT& op)
+{
+    if (vdb.hasGrid()) {
+        return vdb.getGrid().apply<GridTypeListT>(op);
+    }
+    return false;
+}
+
+/// @brief If the given primitive's grid resolves to one of the listed grid types,
+/// invoke the functor @a op on the resolved grid.
+/// @return @c true if the functor was invoked, @c false otherwise
+/// @details If @a makeUnique is true, deep copy the grid's tree before
+/// invoking the functor if the tree is shared with other grids.
+template<typename GridTypeListT, typename OpT>
+inline bool
+GEOvdbApply(GEO_PrimVDB& vdb, OpT& op, bool makeUnique = true)
+{
+    if (vdb.hasGrid()) {
+        if (!makeUnique) {
+            return vdb.getGrid().apply<GridTypeListT>(op);
+        }
+        internal::MakeUniqueOp<OpT> uniqueOp(op);
+        return vdb.getGrid().apply<GridTypeListT>(uniqueOp);
```

Comment:
Instead of having `MakeUniqueOp` rely on the entire functor, can it not be a separate process whose only job is to copy the grid? i.e.
```
if (vdb.hasGrid()) {
    if (makeUnique) {
        internal::MakeUniqueOp uniqueOp;
        vdb.getGrid().apply<GridTypeListT>(uniqueOp);
    }
    return vdb.getGrid().apply<GridTypeListT>(op);
}
```

Reply:
That's a good question.  I wanted to avoid iterating through the type list twice, since the type checks involve string comparisons, but that overhead is of course negligible compared to the cost of copying the tree when the grid is not unique.  So perhaps the best solution is to check for uniqueness before applying `MakeUniqueOp`: 
```
if (vdb.hasGrid()) {
    auto gridPtr = vdb.getGridPtr();
    if (makeUnique) {
        auto treePtr = gridPtr->baseTreePtr();
        if (treePtr.use_count() > 2) { // grid + treePtr = 2
            internal::MakeUniqueOp uniqueOp;
            gridPtr->apply<GridTypeListT>(uniqueOp);
        }
    }
    return gridPtr->apply<GridTypeListT>(op);
}

```
Then `MakeUniqueOp` can be simplified to
```
struct MakeUniqueOp
{
    template<typename GridT>
    void operator()(GridT& grid)
    {
        if (auto treePtr = grid.constTreePtr()) {
            grid.setTree(treePtr->copy());
        }
    }
};
```
Does that seem reasonable?

Reply:
Yeah this looks good to me!